metadata/src/main/java/org/apache/kafka/controller/PartitionChangeBuilder.java

```diff
@@ -104,53 +117,73 @@ public PartitionChangeBuilder setTargetAdding(List<Integer> targetAdding) {
-                }
-            }
-            if (uncleanElectionOk.get()) {
-                for (int replica : targetReplicas) {
-                    if (isAcceptableLeader.apply(replica)) {
-                        this.node = replica;
-                        this.unclean = true;
-                        return;
-                    }
-                }
+        // 2. Attempt preferred replica election
+        int preferredReplica = targetReplicas.get(0);
+        if (targetIsr.contains(preferredReplica) && isAcceptableLeader.apply(preferredReplica)) {
+            return new ElectionResult(preferredReplica, false);
+        }
+
+        // 3. Preferred replica was not elected, only continue if the current leader is not a valid leader
+        if (targetIsr.contains(partition.leader) && isAcceptableLeader.apply(partition.leader)) {
+            // Don't consider a new leader since the current leader meets all the constraints
+            return new ElectionResult(partition.leader, false);
+        }
+
+        // 4. Attempt to keep the partition online based on the ISR
+        Optional<Integer> bestLeader = targetReplicas.stream()
+            .skip(1)
+            .filter(replica -> targetIsr.contains(replica) && isAcceptableLeader.apply(replica))
+            .findFirst();
+        if (bestLeader.isPresent()) {
+            return new ElectionResult(bestLeader.get(), false);
+        }
+
+        if (election == Election.UNCLEAN) {
+            // 5. Attempt unclean leader election
+            Optional<Integer> uncleanLeader = targetReplicas.stream()
+                .filter(replica -> isAcceptableLeader.apply(replica))
+                .findFirst();
+            if (uncleanLeader.isPresent()) {
+                return new ElectionResult(uncleanLeader.get(), true);
             }
-            this.node = NO_LEADER;
-            this.unclean = false;
         }
+
+        return new ElectionResult(NO_LEADER, false);
     }
 
     private void tryElection(PartitionChangeRecord record) {
-        BestLeader bestLeader = new BestLeader();
-        if (bestLeader.node != partition.leader) {
-            log.debug("Setting new leader for topicId {}, partition {} to {}", topicId, partitionId, bestLeader.node);
-            record.setLeader(bestLeader.node);
-            if (bestLeader.unclean) {
+        ElectionResult electionResult = electLeader();
+        if (electionResult.node != partition.leader) {
+            log.debug("Setting new leader for topicId {}, partition {} to {}", topicId, partitionId, electionResult.node);
+            record.setLeader(electionResult.node);
+            if (electionResult.unclean) {
                 // If the election was unclean, we have to forcibly set the ISR to just the
                 // new leader. This can result in data loss!
-                record.setIsr(Collections.singletonList(bestLeader.node));
+                record.setIsr(Collections.singletonList(electionResult.node));
```

Comment:
This is also an existing issue. We set the ISR here, but it can be overridden to targetIsr in tobuild() later.

Reply:
You are right! Let me add a test to confirm this.

Reply:
Fixed this and added tests for it. Confirmed that the tests fail without these changes.